### PR TITLE
[Implementación Swagger] - Implementation swagger

### PR DESCRIPTION
**Pull request by AleFalcon**
## Summary
Implementatio Swagger for documentation of the API

## Trello Card
https://trello.com/c/56TiirpO/40-swagger

### CODE REVIEW

File: src/main/java/wolox/training/controllers/BookApiController.java [hunk 1]
```diff
@@ -29,6 +32,7 @@ public String greeting(@RequestParam(name="name", required=false, defaultValue="
     }
 
     @PostMapping
+    @ApiOperation(value = "Given a book, create one, and return the book.", response = Book.class)
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/BookApiController.java:37]: You could add @ApiResponses to register the possible httpStatus of each controller.

File: src/main/java/wolox/training/controllers/BookApiController.java [hunk 2]
```diff
@@ -37,6 +41,7 @@ public Book create(@RequestBody Book book, Model model) {
     }
 
     @PutMapping
+    @ApiOperation(value = "Given the isbn of the book, you can update the book, and return the book.", response = Book.class)
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/BookApiController.java:52]: The same as previously commented.

File: src/main/java/wolox/training/controllers/BookApiController.java [hunk 3]
```diff
@@ -45,11 +50,13 @@ public Book update(@RequestBody Book book) {
     }
 
     @DeleteMapping
+    @ApiOperation(value = "Given the isbn of the book, you delete the book, and return void", response = void.class)
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/BookApiController.java:68]: The same as previously commented

File: src/main/java/wolox/training/controllers/BookApiController.java [hunk 4]
```diff
@@ -45,11 +50,13 @@ public Book update(@RequestBody Book book) {
...
     public void delete(@RequestParam(name="isbn", required=true) String isbn) {
         bookRepository.delete(found_book(isbn, bookRepository));
     }
 
     @GetMapping("{isbn}")
+    @ApiOperation(value = "Given the isbn of the book, return the book asked", response = Book.class)
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/BookApiController.java:81]: The same as previously commented

File: src/main/java/wolox/training/controllers/UserApiController.java [hunk 5]
```diff
@@ -19,13 +20,16 @@
 
     @PostMapping
     @ResponseStatus(HttpStatus.CREATED)
+    @ApiOperation(value = "Given a user, create one, and return the user.", response = User.class)
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/UserApiController.java:31]: You could add @ApiResponses to register the possible httpStatus of each controller.

File: src/main/java/wolox/training/controllers/UserApiController.java [hunk 6]
```diff
@@ -19,13 +20,16 @@
...
     public User create(@RequestBody User user) {
         User new_user = new User(user.getName(), user.getUsername(), user.getBirthday());
         userRepository.save(new_user);
         return new_user;
     }
 
+
     @PutMapping
+    @ApiOperation(value = "Given the username of a user, you can update the user, and return the user.", response = User.class)
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/UserApiController.java:40]: The same as previously commented

File: src/main/java/wolox/training/controllers/UserApiController.java [hunk 7]
```diff
@@ -35,21 +39,25 @@ public User update(@RequestBody User user) {
     }
 
     @DeleteMapping
+    @ApiOperation(value = "Given the username of a user, you delete the user, and return void", response = void.class)
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/UserApiController.java:57]: The same as previously commented

File: src/main/java/wolox/training/controllers/UserApiController.java [hunk 8]
```diff
@@ -35,21 +39,25 @@ public User update(@RequestBody User user) {
...
     public void delete(@RequestParam(name="username", required=true) String username) {
         userRepository.delete(found_user(username, userRepository));
     }
 
     @GetMapping("{username}")
+    @ApiOperation(value = "Given the username of a user, return the user asked", response = User.class)
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/UserApiController.java:70]: The same as previously commented

File: src/main/java/wolox/training/controllers/UserApiController.java [hunk 9]
```diff
@@ -35,21 +39,25 @@ public User update(@RequestBody User user) {
...
     public User read(@PathVariable String username) {
         return found_user(username, userRepository);
     }
 
     @GetMapping("/list_books/{username}")
+    @ApiOperation(value = "Given the username of a user, return a book list of the user", response = Book.class)
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/UserApiController.java:91]: The same as previously commented

File: src/main/java/wolox/training/controllers/UserApiController.java [hunk 10]
```diff
@@ -35,21 +39,25 @@ public User update(@RequestBody User user) {
...
     public List<Book> list_book(@PathVariable String username) {
         return found_user(username, userRepository).getListBooks();
     }
 
     @PutMapping("/add_book")
+    @ApiOperation(value = "Given the username of a user and the isbn of a book, the book is added to user, return the user", response = void.class)
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/UserApiController.java:97]: The same as previously commented

File: src/main/java/wolox/training/models/Book.java [hunk 11]
```diff
@@ -12,18 +14,22 @@
...
 @Entity
 @Table(name="books")
+@ApiModel(description = "Books from my API")
 public class Book {
     @Id
     @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "BOOK_SEQ")
     @SequenceGenerator(name = "BOOK_SEQ", sequenceName = "BOOK_SEQ")
     private long id;
 
+    @ApiModelProperty(notes = "The book genre: could be horror, comedy, drama, etc")
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: You can add the examples as follows:
`@ApiModelProperty(notes = "genre", example = "Horror")`

And in the mandatory fields you can add the following:
`@ApiModelProperty(notes = "genre", required = true, example = "Horror")`

File: src/main/java/wolox/training/models/User.java [hunk 12]
```diff
@@ -36,7 +36,6 @@
     private LocalDate birthday;
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/User.java:43]: Add swagger to user model

File: src/main/java/wolox/training/models/Book.java [hunk 13]
```diff
@@ -12,18 +14,22 @@
...
     @Column(nullable = false)
     private String genre;
 
+    @ApiModelProperty(notes = "The book author: this is the author of the book")
     @Column(nullable = false)
     private String author;
 
+    @ApiModelProperty(notes = "The book image: this is the image the book")
     @Column(nullable = false)
     private String image;
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java:39]: Add swagger to the rest of the fields of the book model

**Review by Gianni-Nicolas — CHANGES_REQUESTED** [review]
Great Job

**Review by Gianni-Nicolas — DISMISSED** [review]

**Review by Gianni-Nicolas — DISMISSED** [review]

**Review by Gianni-Nicolas — DISMISSED** [review]

File: src/main/java/wolox/training/utils/ErrorConstants.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
+package wolox.training.utils;
+
+public class ErrorConstants {
+
+  //PRECONDITIONS BOOK
+  public static String NOT_NULL_GENRE = "The genre field cannot be null.";
+  public static String NOT_EMPTY_GENRE = "The genre field cannot be empty.";
```
Comment (facucompe) [inline src/main/java/wolox/training/utils/ErrorConstants.java]: You can implement a generic string here. Something like this :
`The %s field cannot be empty`
And then do:
`String.format(ErrorConstants.fieldEmpty, "genre")`

**Review by facucompe — CHANGES_REQUESTED** [review]

**Review by AleFalcon — COMMENTED** [review]
Okay

**Review by Gianni-Nicolas — APPROVED** [review]

**Review by facucompe — APPROVED** [review]

**Pull request merged**